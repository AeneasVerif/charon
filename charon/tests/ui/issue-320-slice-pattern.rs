fn slice_pat1() {
    let array: [_; 4] = [0; 4];
    let [_a, _b @ .., _c] = array;
}

fn slice_pat2() {
    let array_ref: &[_; 4] = &[0; 4];
    let [_a, _b @ .., _c] = array_ref;
}

fn slice_pat3() {
    let slice: &[_] = &[0; 4];
    let [_a, _b @ .., _c] = slice else { panic!() };
}
