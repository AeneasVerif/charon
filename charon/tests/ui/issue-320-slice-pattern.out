# Final LLBC before serialization:

// Full name: test_crate::slice_pat1
fn slice_pat1()
{
    let @0: (); // return
    let array@1: Array<i32, 4 : usize>; // local
    let _a@2: i32; // local
    let _b@3: Array<i32, 2 : usize>; // local
    let _c@4: i32; // local
    let @5: &'_ (Array<i32, 4 : usize>); // anonymous local
    let @6: &'_ (i32); // anonymous local
    let @7: &'_ (Array<i32, 4 : usize>); // anonymous local
    let @8: &'_ (Slice<i32>); // anonymous local
    let @9: &'_ (Array<i32, 4 : usize>); // anonymous local
    let @10: &'_ (i32); // anonymous local

    storage_live(array@1)
    array@1 := @ArrayRepeat<'_, i32, 4 : usize>(const (0 : i32))
    storage_live(_a@2)
    storage_live(@9)
    @9 := &array@1
    storage_live(@10)
    @10 := @ArrayIndexShared<'_, i32, 4 : usize>(move (@9), const (0 : usize))
    _a@2 := copy (*(@10))
    storage_live(_b@3)
    storage_live(@7)
    @7 := &array@1
    storage_live(@8)
    @8 := @ArraySubSliceShared<'_, i32, 4 : usize>(move (@7), const (1 : usize), const (3 : usize))
    _b@3 := copy (*(@8))
    storage_live(_c@4)
    storage_live(@5)
    @5 := &array@1
    storage_live(@6)
    @6 := @ArrayIndexShared<'_, i32, 4 : usize>(move (@5), const (3 : usize))
    _c@4 := copy (*(@6))
    @0 := ()
    storage_dead(_c@4)
    storage_dead(_b@3)
    storage_dead(_a@2)
    storage_dead(array@1)
    @0 := ()
    return
}

// Full name: test_crate::slice_pat2
fn slice_pat2()
{
    let @0: (); // return
    let array_ref@1: &'_ (Array<i32, 4 : usize>); // local
    let @2: &'_ (Array<i32, 4 : usize>); // anonymous local
    let @3: Array<i32, 4 : usize>; // anonymous local
    let _a@4: &'_ (i32); // local
    let _b@5: &'_ (Array<i32, 2 : usize>); // local
    let _c@6: &'_ (i32); // local
    let @7: &'_ (Array<i32, 4 : usize>); // anonymous local
    let @8: &'_ (i32); // anonymous local
    let @9: &'_ (Array<i32, 4 : usize>); // anonymous local
    let @10: &'_ (Slice<i32>); // anonymous local
    let @11: &'_ (Array<i32, 4 : usize>); // anonymous local
    let @12: &'_ (i32); // anonymous local

    storage_live(array_ref@1)
    storage_live(@2)
    storage_live(@3)
    @3 := @ArrayRepeat<'_, i32, 4 : usize>(const (0 : i32))
    @2 := &@3
    array_ref@1 := &*(@2)
    storage_dead(@2)
    storage_live(_a@4)
    storage_live(@11)
    @11 := &*(array_ref@1)
    storage_live(@12)
    @12 := @ArrayIndexShared<'_, i32, 4 : usize>(move (@11), const (0 : usize))
    _a@4 := &*(@12)
    storage_live(_b@5)
    storage_live(@9)
    @9 := &*(array_ref@1)
    storage_live(@10)
    @10 := @ArraySubSliceShared<'_, i32, 4 : usize>(move (@9), const (1 : usize), const (3 : usize))
    _b@5 := &*(@10)
    storage_live(_c@6)
    storage_live(@7)
    @7 := &*(array_ref@1)
    storage_live(@8)
    @8 := @ArrayIndexShared<'_, i32, 4 : usize>(move (@7), const (3 : usize))
    _c@6 := &*(@8)
    @0 := ()
    storage_dead(_c@6)
    storage_dead(_b@5)
    storage_dead(_a@4)
    storage_dead(@3)
    storage_dead(array_ref@1)
    @0 := ()
    return
}

// Full name: test_crate::slice_pat3
fn slice_pat3()
{
    let @0: (); // return
    let slice@1: &'_ (Slice<i32>); // local
    let @2: &'_ (Array<i32, 4 : usize>); // anonymous local
    let @3: &'_ (Array<i32, 4 : usize>); // anonymous local
    let @4: Array<i32, 4 : usize>; // anonymous local
    let _a@5: &'_ (i32); // local
    let _b@6: &'_ (Slice<i32>); // local
    let _c@7: &'_ (i32); // local
    let @8: usize; // anonymous local
    let @9: usize; // anonymous local
    let @10: bool; // anonymous local
    let @11: &'_ (Slice<i32>); // anonymous local
    let @12: usize; // anonymous local
    let @13: usize; // anonymous local
    let @14: &'_ (i32); // anonymous local
    let @15: &'_ (Slice<i32>); // anonymous local
    let @16: usize; // anonymous local
    let @17: usize; // anonymous local
    let @18: &'_ (Slice<i32>); // anonymous local
    let @19: &'_ (Slice<i32>); // anonymous local
    let @20: &'_ (i32); // anonymous local

    storage_live(@8)
    storage_live(@9)
    storage_live(@10)
    storage_live(slice@1)
    storage_live(@2)
    storage_live(@3)
    storage_live(@4)
    @4 := @ArrayRepeat<'_, i32, 4 : usize>(const (0 : i32))
    @3 := &@4
    @2 := &*(@3)
    slice@1 := @ArrayToSliceShared<'_, i32, 4 : usize>(move (@2))
    storage_dead(@2)
    storage_dead(@3)
    storage_live(_a@5)
    storage_live(_b@6)
    storage_live(_c@7)
    @8 := len(*(slice@1))
    @9 := const (2 : usize)
    @10 := move (@8) >= move (@9)
    if move (@10) {
    }
    else {
        storage_dead(_c@7)
        storage_dead(_b@6)
        storage_dead(_a@5)
        panic(core::panicking::panic_explicit)
    }
    storage_live(@19)
    @19 := &*(slice@1)
    storage_live(@20)
    @20 := @SliceIndexShared<'_, i32>(move (@19), const (0 : usize))
    _a@5 := &*(@20)
    storage_live(@15)
    @15 := &*(slice@1)
    storage_live(@16)
    @16 := len(*(slice@1))
    storage_live(@17)
    @17 := copy (@16) ub.- const (1 : usize)
    storage_dead(@16)
    storage_live(@18)
    @18 := @SliceSubSliceShared<'_, i32>(move (@15), const (1 : usize), copy (@17))
    _b@6 := &*(@18)
    storage_live(@11)
    @11 := &*(slice@1)
    storage_live(@12)
    @12 := len(*(slice@1))
    storage_live(@13)
    @13 := copy (@12) ub.- const (1 : usize)
    storage_dead(@12)
    storage_live(@14)
    @14 := @SliceIndexShared<'_, i32>(move (@11), copy (@13))
    _c@7 := &*(@14)
    @0 := ()
    storage_dead(_c@7)
    storage_dead(_b@6)
    storage_dead(_a@5)
    storage_dead(@4)
    storage_dead(slice@1)
    @0 := ()
    return
}



